src/tree/newtree/provisional_tree.rs

```diff
@@ -0,0 +1,162 @@
+    /// `original_tree.node_index`. If the `original_tree` is `None`, it has to
+    /// be provided when creating the provisional tree.
+    own_node_index: NodeIndex,
+    /// The provisional public tree. It's either the base for a new
+    /// `RatchetTree` or the staging ground for changes to an already existing
+    /// `RatchetTree`.
+    provisional_public_tree: BinaryTree<Option<Node>>,
+
+    /// The provisional private key. If the `ProvisionalTree` is created from a
+    /// `KeyPackageBundle`, either to create a fresh group or from a `Welcome`
+    /// message, it should be `Some`. If the `ProvisionalTree` is created from a
+    /// `RatchetTreeBundle`, it is `None` initially and might be set later if
+    /// the owner's leaf key is updated.
+    provisional_private_key: Option<HPKEPrivateKey>,
+
+    /// The `provisional_path_keys` are always initialized as an empty `HashMap`
+    /// and populated either in the process of creating the `ProvisionalTree` or
+    /// when applying commits.
+    provisional_path_keys: HashMap<NodeIndex, HPKEPrivateKey>,
+}
+
+#[allow(dead_code, unused_variables)]
+impl<'a> ProvisionalTree<'a> {
+    /// Create a `ProvisionalTree` from a given `RatchetTreeBundle` reference.
+    /// This should happen in the process of creating or applying a commit.
+    pub(crate) fn from_ratchet_tree_bundle(original_tree: &'a RatchetTreeBundle) -> Self {
+        let mut tree = Vec::new();
+        tree.resize(original_tree.public_tree.size(), None);
+        let provisional_public_tree = BinaryTree::from(tree);
+        let provisional_private_key = None;
+        let provisional_path_keys = HashMap::new();
+        ProvisionalTree {
+            original_tree: Some(original_tree),
+            own_node_index: original_tree.node_index,
+            provisional_public_tree,
+            provisional_private_key,
+            provisional_path_keys,
+        }
+    }
+
+    /// Create a `ProvisionalTree` from a `KeyPackageBundle`. This function is
+    /// used when creating a new group.
+    pub(crate) fn from_key_package_bundle(key_package_bundle: KeyPackageBundle) -> Self {
+        let node_contents = Some(NodeContents::LeafContents(key_package_bundle.key_package));
+        let node = Node::new_leaf(node_contents);
+        let provisional_public_tree = BinaryTree::from(vec![Some(node)]);
+        let provisional_private_key = Some(key_package_bundle.private_key);
+        let provisional_path_keys = HashMap::new();
+        let node_index = NodeIndex::from(0u32);
+        ProvisionalTree {
+            original_tree: None,
+            own_node_index: node_index,
+            provisional_public_tree,
+            provisional_private_key,
+            provisional_path_keys,
+        }
+    }
+
+    /// Create a `ProvisionalTree` from a `KeyPackageBundle` and some
+    /// information from a `Welcome` message.
+    pub(crate) fn from_welcome_info(
```

Comment:
Same question as above. Maybe it would be good to add some explanation what the exact prupose of a provisional tree is. So far we only discussed in the context of `create_commit` and `apply_commit`.